<a href="https://colab.research.google.com/github/sebassanchez5680-tech/Prueba/blob/main/Actividad_Calificada_07_Datos_Duplicados_271757.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** | |
| **Matrícula** | |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [123]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [124]:
dup_first = hoteles.duplicated(keep='first').sum() #contamos los repetidos unicamente conservando como false la primera aparicion

dup_last = hoteles.duplicated(keep='last').sum() #similar a first pero en este caso la ultima aparicon se le asgina el false

Todo = hoteles.duplicated(keep=False).sum() #aqui contamos a todas por igual, entonces si un valor se repite se marca todas las repeticiones

print(f"Filas keep='first': {dup_first}")
print(f"Filas keep='last': {dup_last}")
print(f"Filas keep=False: {Todo}")

Filas keep='first': 31994
Filas keep='last': 31994
Filas keep=False: 40165


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [125]:
proporcion = hoteles.duplicated().mean() #calculamos el porcentaje de sobrantes respecto del total del dataset
conteo_nombre_edad = hoteles.value_counts() #contamos cuantas veces aparecen cada uno de los valores
repetidos = conteo_nombre_edad[conteo_nombre_edad > 1] # contamos a los grupos que aparecen mas de 1 vez

print(f'Porcentaje de filas sobrantes: {proporcion * 100:.2f} %')
print('Grupos repetidos: ', len(repetidos))

Porcentaje de filas sobrantes: 26.80 %
Grupos repetidos:  17


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [126]:

hoteles_diag = hoteles.copy() #copia

hoteles_diag['es_duplicado'] = hoteles_diag.duplicated(keep='first') #booleano que identifica las repeticiones posteriores a la primera aparicion

# Agrupamos por hotel para sumar los duplicados y calcular la proporcion (media)
# Con agg aplicamos la operacion para obetener el promedio
duplicados_por_hotel = hoteles_diag.groupby('hotel')['es_duplicado'].agg( Sobrantes='sum',Porcentaje=lambda x: (x.mean() * 100).round(2))

print(duplicados_por_hotel)
print(len(hoteles))

              Sobrantes  Porcentaje
hotel                              
City Hotel        25902       32.65
Resort Hotel       6092       15.21
119390


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

En general el dataset contiene 119,390 reservas y de esas 31,994 son filas sobrantes, representando el 26.80% del total. practicamente un cuarto de todo el data set son datos repetidos y esto es malo.
Es suficientemente grande como para cambiar los conteos y metricas.
A nivel de los hoteles tenemos que City Hotel sus duplicados son de 32.65% mientras que Resort Hotel  es de 15.21%, en city hotel sus duplicados representan mas del doble de los que se tienen en resort hotel afectando claramente mas a la obtencion de conteos y metricas

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [127]:

mascara = hoteles.duplicated(keep=False)
columnas_vista = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month', # filtramos todas las copias de los grupos repetidos
                  'arrival_date_day_of_month', 'adults', 'country', 'market_segment',
                  'deposit_type', 'adr', 'reservation_status']

# Ordenamos por hotel, fecha de llegada y tiempo de antelacion asi agrupamos las reservas identicas, el tiempo es un buen ordenante para ver cronologicamente las filas e identificar si hay alguna que aun
hoteles[mascara].sort_values(by=['hotel', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'lead_time'])[columnas_vista].head(12)

,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adults,country,market_segment,deposit_type,adr,reservation_status
40574,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40580,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40588,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40562,City Hotel,117,2015,August,1,1,ITA,Offline TA/TO,No Deposit,72.5,Check-Out
40563,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40564,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40566,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Check-Out
40567,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Check-Out
40568,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40569,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [128]:
conteo = hoteles.value_counts(dropna=False) # con dropna=False hacemmos que las combinaciones con NaN se cuenten tambien
grupos = conteo[conteo > 1] #para ser considerado como un grupo se tiene que tener mas de 1 valor

print("10 Grupos mas comunes:")
print(grupos.value_counts().head(10))

print("Grupo más grande:")
print(grupos.head(1))

10 Grupos mas comunes:
count
2     5034
3      969
4      497
5      275
6      171
10     123
7      115
8      103
20      97
9       93
Name: count, dtype: int64
Grupo más grande:
hotel       is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type  adr    required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel  1            277        2016               November            46                        7                          1                        2                     2       0.0       0       BB    PRT      Groups          TA/TO                 0  

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [129]:
# Filtramos las filas que están involucradas en algún grupo repetido.
duplicados_involucrados = hoteles[hoteles.duplicated(keep=False)].copy()

# Comparamos las distribuciones como proporciones
market_total = hoteles['market_segment'].value_counts(normalize=True, dropna=False) #contamos los valores de market
market_dup = duplicados_involucrados['market_segment'].value_counts(normalize=True, dropna=False) #contamos los duplicados

deposit_total = hoteles['deposit_type'].value_counts(normalize=True, dropna=False) #hacemos lo mismo para deposit
deposit_dup = duplicados_involucrados['deposit_type'].value_counts(normalize=True, dropna=False)

tasa_cancelacion_total = hoteles['is_canceled'].mean() #calculamos las tasas de desocupacion
tasa_cancelacion_dup = duplicados_involucrados['is_canceled'].mean()

print("Distribución de market_segment:")
print(pd.concat([market_total.rename('total'), market_dup.rename('duplicados')], axis=1))

print("\nDistribución de deposit_type:")
print(pd.concat([deposit_total.rename('total'), deposit_dup.rename('duplicados')], axis=1))

print(f"\nTasa de cancelación - dataset completo: ({tasa_cancelacion_total*100:.2f}%)")
print(f"Tasa de cancelación - filas involucradas en duplicados: ({tasa_cancelacion_dup*100:.2f}%)")

Distribución de market_segment:
                   total  duplicados
market_segment                      
Online TA       0.473046    0.205876
Offline TA/TO   0.202856    0.303996
Groups          0.165935    0.416109
Direct          0.105587    0.034981
Corporate       0.044350    0.036699
Complementary   0.006223    0.001867
Aviation        0.001985    0.000473
Undefined       0.000017         NaN

Distribución de deposit_type:
                 total  duplicados
deposit_type                      
No Deposit    0.876464    0.641678
Non Refund    0.122179    0.356430
Refundable    0.001357    0.001892

Tasa de cancelación - dataset completo: (37.04%)
Tasa de cancelación - filas involucradas en duplicados: (58.39%)


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

a)  Con la informacion que se tiene no es posible demostrar que todas representen errores de registro. Nos hace falta algo asi como un identificador unico por reserva. Asi podriamos determinar si de verdad se repite.

b) duplicated() considera los NaN iguales entre si cuando comparan las filas. Al momento de que dos filas tienen NaN en agent y company y coinciden en las otras columnas, se marcan como duplicadas.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [130]:
# Eliminamos duplicados exactos y renumeramos el indice para mantener consistencia (keep='first')
hoteles_limpio = hoteles.drop_duplicates(keep='first', ignore_index=True)

print(f"Filas originales: {len(hoteles)}")
print(f"Filas después de limpieza: {len(hoteles_limpio)}")
print(f"Filas eliminadas: {len(hoteles) - len(hoteles_limpio)}")

Filas originales: 119390
Filas después de limpieza: 87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [131]:
#metricas para el dataset original.
resumen_original = (hoteles.groupby('hotel').agg(reservas=('hotel', 'size'),tasa_cancelacion=('is_canceled', 'mean'),adr_promedio=('adr', 'mean')).reset_index())

# metricas sin duplicados.
resumen_limpio = (hoteles_limpio.groupby('hotel').agg(reservas=('hotel', 'size'),tasa_cancelacion=('is_canceled', 'mean'),adr_promedio=('adr', 'mean')).reset_index())

print("Resumen original  :")
print(resumen_original)

print("Resumen sin duplicados  :")
print(resumen_limpio)

Resumen original  :
          hotel  reservas  tasa_cancelacion  adr_promedio
0    City Hotel     79330          0.417270    105.304465
1  Resort Hotel     40060          0.277634     94.952930
Resumen sin duplicados  :
          hotel  reservas  tasa_cancelacion  adr_promedio
0    City Hotel     53428          0.300386    110.985944
1  Resort Hotel     33968          0.234809     99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [132]:
# ponemos una etiqueta para distinguir las versiones.
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

# unimos los dos conjuntos con pd.concat().
comparacion = (pd.concat([resumen_original, resumen_limpio], ignore_index=True).sort_values(by='hotel').reset_index(drop=True))

print(comparacion)

          hotel  reservas  tasa_cancelacion  adr_promedio         version
0    City Hotel     79330          0.417270    105.304465        original
1    City Hotel     53428          0.300386    110.985944  sin duplicados
2  Resort Hotel     40060          0.277634     94.952930        original
3  Resort Hotel     33968          0.234809     99.025346  sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

a) La metrica que mas cambia es la tasa de cancelacion, aunque veamos que se mueven apenas unas decimas, recordemos que se trata de una tasa, por lo cual esos pequeños movimientos son significantes.

El hotel que mayor cambio tiene es el city hotel precisamente por el hecho de que fue el que mas reservas perdio al momenot de quitar duplicados, por lo que los cambios de sus metricas son mayores

b) Ocasionaria que el modelo de mas peso a los valores repetidos, como por ejemplo el dia donde mas se reserva, haciendo que los valores predichos esten inflados por las repeticiones

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [133]:
duplicados_restantes = hoteles_limpio.duplicated().sum() #contamos si hay duplicado

# Verificamos que ningún índice aparezca más de una vez.
indices_repetidos = hoteles_limpio.index.duplicated().sum() #verificamos que no aparezcan mas de 1 vez los indices contandolos

indice_correcto = hoteles_limpio.index.equals(pd.RangeIndex(start=0, stop=len(hoteles_limpio))) #comprobamos que el indice sea exactamente 0, 1, 2, ...

print("Duplicados exactos restantes:", duplicados_restantes)
print("Indices repetidos:", indices_repetidos)
print("Indice renumerado correctamente:", indice_correcto)

Duplicados exactos restantes: 0
Indices repetidos: 0
Indice renumerado correctamente: True


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [134]:
# quitamos la fecha de estado para buscar reservas que solo difieren en el campo
columnas_sin_fecha_estado = hoteles_limpio.columns.drop('reservation_status_date')

sobrantesp = hoteles_limpio.duplicated(subset=columnas_sin_fecha_estado).sum()#contamos las filas sobrantes con el nuevo criterio
print(f"Sobrantes parciales: {sobrantesp}")

mascara_parcial = hoteles_limpio.duplicated(subset=columnas_sin_fecha_estado, keep=False) #buscamos los duplicados
columnas_revision = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month','arrival_date_day_of_month', 'adr', 'reservation_status', 'reservation_status_date'] #lista de nombres de columnas que queremos mostrar

print("\nDistribución de estado en duplicados parciales:")
print(hoteles_limpio[mascara_parcial]['reservation_status'].value_counts())

Sobrantes parciales: 245

Distribución de estado en duplicados parciales:
reservation_status
Canceled    471
Name: count, dtype: int64


**Respuesta 4.2:**

Las filas detectadas coinciden en todas las columnas menos en reservation_status_date. Esto nos puede representar una misma reserva cuyo estado fue actualizado en diferentes fechas, por lo cual puede ser informacion util si se evalua mas a fonfo las reservaciones, por lo cual considero que eliminarlas no seria conveniente.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

Criterio de duplicado: se considero duplicado exacto un registro el cual sus 32 atributos coinciden exactamente con otra fila.

Regla de conservación: se uso keep='first', conservando la primera aparicion de cada fila. En este caso no es muy determinante ya que tanto keep='first' o keep='last' producen el mismo contenido porque las filas son iguales.

Resultado: Originalmente se contaba con alrededor de 119,390 filas las cuales pasaron a ser 87,396 filas, eliminandose asi 31,994 filas.

Supuestos y limitaciones: se tomo que las filas repetidas era un error por lo que se quitaron. Sin embargo, el dataset no contiene un identificador de reserva, por lo que no se puede demostrar que cada duplicado corresponda a un error de captura como tal. Por ello, los duplicados parciales basados en todas las columnas excepto reservation_status_date se conservaron y se analizaron por separado.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

1. Situación en la que no convendria eliminar duplicados exactos

No convendria eliminarlos si el dataset representa eventos o reservas independientes y sin un id. Dos clientes o dos reservas diferentes pueden tener exactamente los mismos atributos registrados. En este caso, eliminar una fila podria significar eliminar un registro valido.

2. Información adicional que pediría al proveedor
Lo primordial seria contar con un identificador unico de reserva. Tambien seria util contar con fecha/hora de creacion de la reserva, fecha/hora de actualizacion. Estas variables permitirian distinguir entre una copia o varias reservas reales con caracteristicas identicas.

3. Por qué duplicated().sum() == 0 no garantiza que el dataset esté libre de duplicados
Porque duplicated() solo detecta coincidencias de acuerdo a las columnas que se estan comparando. Dos registros de la misma reserva pueden diferir en una columna y no serian duplicados exactos.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.